src/core/Target.mli

```diff
@@ -0,0 +1,146 @@
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+(** Types for describing targets.
+
+    See also {!Input_to_core_t}, which has a similar set of types used when
+    pysemgrep generates targets that have slightly less information (e.g.,
+    these types have expanded information about the targets' locations). *)
+
+(** The location of a Semgrep target. This contains all of the target-specific
+    details needed to be able to determine how to scan a target, e.g.,
+
+    {ul
+      {- What products should we select rules from?}
+      {- Where can we get the contents of the target?}
+      {- What language should we analyze the target as?}
+    }
+
+    However, it does not contain the actual contents (parsed or otherwise) of
+    the target itself.
+ *)
+type t = Code of code | Lockfile of lockfile [@@deriving show]
+
+and code = {
+  path : path;
+  analyzer : Xlang.t;  (** The analyzer to use when scanning this target. *)
+  products : Semgrep_output_v1_t.product list;
+      (** The products which should scan this target. This is used for
+          selecting the relevant set of rules. *)
+  lockfile : lockfile option;
+      (** Optional lockfile associated with this target.
+
+          The association is namely that this target has its dependencies
+          specified by the given lockfile. Core doesn't need to worry about
+          determining these associations; rather, the target selection and
+          generation process must resolve these connections as part of
+          generating code targets. *)
+}
+[@@deriving show]
+(** A "normal" semgrep target, comprising source code (or, for
+   regex/generic, arbitrary text data) to be executed. See also {!Xtarget.t},
+   an augmented version which also has the contents. *)
+
+and lockfile = {
+  path : path;
+  kind : Lockfile_kind.t;
+      (** The type of lockfile this is. Analogous to analyzer for a source code
+          target. *)
+  manifest : manifest option;
+      (** Optionally, a manifest file associated with this lockfile. *)
+}
+[@@deriving show]
+(** A lockfile to be used during matching. See also {!Lockfile_xtarget.t}, an
+    augmented version with the contents of the lockfile. *)
+
+and manifest = {
+  path : path;
+  kind : Manifest_kind.t;
+      (** The type of manifest this is. Analogous to analyzer for a source code
+        target. *)
+}
+[@@deriving show]
+(** A manifest file to be used during matching. See also
+    {!Lockfile_xtarget.manifest}, which also has the contents. *)
+
+and path = {
```

Comment:
please avoid mutually recursive construct like 'and' when the type is actually not mutually recursive.
I know sometimes it's nice to put the important type first, and abuse 'and' so that one can order
the types as we please, but better to actually see when types really need to be mutually recursive.

Reply:
I have made this change in 275903281, but I think it is a net-negative. We already have other definitions (e.g., [various taint types for Rule.t](https://github.com/semgrep/semgrep/blob/55412a8d87addd85ec1543c82a4598f4b44847df/src/core/Rule.ml#L168-L304)) defined with `and` which are not mutually recursive, and I think listing the types in what is essentially the reverse order of relevancy simply makes the interface harder to read (further hurt by the fact that you then end up with the most important types sandwiched between other types and functions, not even at the end of the file).